From 9c65faa246edeb70c783f560388f0d0f3d92e587 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Tue, 14 Apr 2020 19:25:41 -0700
Subject: [PATCH] Add explicit cast to invocations of java/nio/Buffer
 functions. Java 9+ added overrides to these functions to return the specific
 subclass, however, when there is a compiler "bug" that when targeting release
 * or below, it will still reference these new methods, causing exceptions at
 runtime on Java 8.


diff --git a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
index 9bbdb53..7a3790f 100644
--- a/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
+++ b/src/org/jetbrains/java/decompiler/modules/decompiler/exps/InvocationExprent.java
@@ -46,6 +46,8 @@ public class InvocationExprent extends Exprent {
 
   private static final BitSet EMPTY_BIT_SET = new BitSet(0);
 
+  private static final VarType JAVA_NIO_BUFFER = new VarType(CodeConstants.TYPE_OBJECT, 0, "java/nio/Buffer");
+
   private String name;
   private String classname;
   private boolean isStatic;
@@ -604,6 +606,12 @@ public class InvocationExprent extends Exprent {
           else if (instance.getPrecedence() > getPrecedence() && !skippedCast) {
             buf.append("(").append(res).append(")");
           }
+          //Java 9+ adds some overrides to java/nio/Buffer's subclasses that alter the return types.
+          //This isn't properly handled by the compiler. So explicit casts are needed to retain J8 compatibility.
+          else if (JAVA_NIO_BUFFER.equals(descriptor.ret) && !JAVA_NIO_BUFFER.equals(rightType)
+              && DecompilerContext.getStructContext().instanceOf(rightType.value, JAVA_NIO_BUFFER.value)) {
+              buf.append("((").append(ExprProcessor.getCastTypeName(JAVA_NIO_BUFFER)).append(")").append(res).append(")");
+          }
           else {
             buf.append(res);
           }
-- 
2.27.0.windows.1

